src/gui/render/graphics_shaders/program_handler.cpp

```diff
@@ -0,0 +1,340 @@
+#include "program_handler.hpp"
+
+#include "gui_render_types.hpp"
+#include "logging.hpp"
+
+#include <GL/glew.h>
+
+#include <filesystem>
+#include <fstream>
+#include <regex>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace gui {
+
+namespace shader {
+
+std::optional<std::string>
+File::get_file_content() {
+    std::ifstream shader_stream(file_, std::ios::in);
+    if (shader_stream.is_open()) {
+        std::stringstream sstr;
+        sstr << shader_stream.rdbuf();
+        std::string shader_code = sstr.str();
+        shader_stream.close();
+        status_ = FileStatus::OK;
+        return shader_code;
+    } else {
+        LOG_ERROR(
+            logging::opengl_logger, "Cannot open {}. Is this the right directory?",
+            file_.string()
+        );
+        status_ = FileStatus::FILE_NOT_FOUND;
+        return {};
+    }
+}
+
+std::string
+get_shader_string(GLuint gl_shader_type) {
+    switch (gl_shader_type) {
+        case GL_VERTEX_SHADER:
+            return "vertex";
+        case GL_FRAGMENT_SHADER:
+            return "fragment";
+        case GL_GEOMETRY_SHADER:
+            return "geometry";
+
+        default:
+            return "NOT A VALID SHADER TYPE";
+    }
+}
+
+void
+Shader::reload() {
+    if (shader_ID_ != 0) {
+        glDeleteShader(shader_ID_);
+    }
+
+    found_uniforms_.clear();
+
+    std::vector<std::string> source_string;
+    std::string shader_type_string = get_shader_string(shader_type_);
+
+    // Create the shader
+    shader_ID_ = glCreateShader(shader_type_);
+
+    for (File file : files_) {
+        std::optional<std::string> file_read = file.get_file_content();
+
+        if (file.get_status() != FileStatus::OK || !file_read) {
+            status_ = ShaderStatus::INVALID_FILE;
+            LOG_ERROR(
+                logging::file_io_logger,
+                "Could not open {}. Check that the files exists.", file.get_file_path()
+            );
+            return;
+        }
+
+        std::string file_text = file_read.value();
+
+        source_string.push_back(file_text);
+
+        std::regex uniform_regex("\n[ |\t]{0,}uniform[ |\t]{1,}(\\w*)[ |\t]{1,}(\\w*);"
```

Comment:
NO NO NO NO NO

`std::regex` is known for being incredibly slow. Is this performance sensitive at all?

Reply:
What is this even for?

Reply:
Well that's why i was checking.